torch_xla/experimental/scan.py

```diff
@@ -82,14 +132,514 @@ def scan(fn, init, xs):
+
+    HloModule Fn, entry_computation_layout={
+      (f32[8], f32[8], f32[8]) -> (f32[8], f32[8])
+    }
+  
+  The HLO computation takes three parameters while `fn` takes two arguments.
+  That's because IR lowering does not distinguish if a leaf data tensor comes from
+  a function argument or from within the function. All data tensors are lowered
+  into HLO parameters. We'll call them "hoisted variables" or `hoisted_vars`, since
+  instead of baking the value of those tensors as literals in the HLO graph,
+  they are turned into additional parameters of the computation.
+  """
+  carry_len = len(init)
+  xs_len = len(xs)
+
+  # Abstractly trace and lower `fn`.
+  # Later we will include `fn_computation` within the while loop body.
+  def make_fake_tensor(v: torch.Tensor) -> torch.Tensor:
+    return torch.empty(
+        v.size(), dtype=v.dtype).to(device).requires_grad_(v.requires_grad)
 
-  for i in range(xs_length):
-    carry, y = fn(carry, tree_map(lambda x: x[i], xs))
-    ys.append(y)
+  device = torch_xla.device()
+  fake_carry = tree_map(make_fake_tensor, init)
+  fake_x = tree_map(lambda v: make_fake_tensor(v[0]), xs)
+  fake_output_carry, fake_output_y = fn(fake_carry, fake_x)
 
-  # Combine the list of PyTrees into one PyTree, where the leaves are
-  # stacked into a new major axis.
-  ys = tree_map(lambda *x: torch.stack(x), *ys)
+  y_len = len(fake_output_y)
+  fn_outputs = fake_output_carry + fake_output_y
+
+  fn_ctx = torch_xla._XLAC.lowering.LoweringContext()
+  fn_ctx.set_name_string("fn_ctx")
+  fn_ctx.build(list(fn_outputs))
+  fn_hlo = fn_ctx.hlo()
+  fn_computation = xb.computation_from_module_proto("fn_computation", fn_hlo)
+
+  # Figure out the shape of `ys` from the abstract tracing.
+  fn_carry_out, fn_y_out = split(fn_outputs, carry_len)
+  assert carry_len + y_len == len(fn_outputs)
+  fn_carry_shapes = [v.shape for v in fn_carry_out]
+  fn_y_shapes = [v.shape for v in fn_y_out]
+  for fn_carry_shape, init_leaf in zip(fn_carry_shapes, init):
+    assert fn_carry_shape == init_leaf.shape, f"`fn` must keep the `carry` shape unchanged. \
+      Got {fn_carry_shape} but expected {init_leaf.shape}"
+
+  builder = Builder('scan')
+  num_iters = next(iter(tree_iter(xs))).size(0)
+  ys = [
+      torch.zeros((num_iters, *fn_y_shape), device=device)
+      for fn_y_shape in fn_y_shapes
+  ]
+  # Start the `curr_iter` loop variable at zero.
+  zero = torch.tensor(0, device=device)
+  builder.add_param(zero)
+
+  # We are building a bigger XLA computation (the while loop) that calls
+  # a smaller computation (`fn_computation`). This is a mapping from
+  # `fn_computation` param ID to While computation param ID.
+  fn_param_id_to_while_param_id: Dict[int, int] = {}
+
+  # Add carry and x.
+  for real, fake in ((init, fake_carry), (xs, fake_x)):
+    for val, fake_val in zip(real, fake):
+      idx = builder.add_param(val)
+      param_id = fn_ctx.tensor_parameter_id(fake_val)
+      if param_id != -1:
```

Comment:
when will it be -1?

Reply:
That is for operations that are not on the device data (OpKindWrapper). I think there is a non-obvious issue/confusion that the mapping gives us all "parameters" for that computation, including inlined constants, but it does not have a parameter ID associated with it for other APIs. Do we have this dependency somewhere to expose these? Those are technically not actual parameters in the computation.

On a similar topic, I wonder that in case of SPMD:
```
    std::vector<xla::Literal> literals =
        runtime::GetComputationClient()->TransferFromDevice(
            UnwrapXlaData(device_data));
```
will be significantly more expensive than having the mapping only traverse ParametersData() and return tensors with `AtenFromXlaTensor` that align with `tensor_parameter_id`, rather than reconstructing a new tensor for all leaf operations. When I tried using sharded tensors with this mapping, it caused a significant synchronous bottleneck - but this can be solved/modified separately.

Reply:
Honestly I don't really understand what OpKindWrapper is so I'm not sure if there's something wrong here. If there is, appreciate advice how I should fix it. I can tell you my high level observation: when lowering LazyTensor IR to HLO, inlined constants and tensors created within the function are hoisted to HLO parameters:

```
    def fn(carry, x):
      foo = torch.zeros(8)
      return carry, x + foo
```

lowers into an HLO like this:

```
    HloModule FnComputation, entry_computation_layout={
      (f32[8], f32[8], f32[8]) -> (f32[8], f32[8])
    }
```

Now when I build the While computation, I must provide some tensor for the third parameter, which is in fact `torch.zeros(8)`. This `tensor_parameter_id` seems to provide that functionality. It also seems to materialize the tensor data on the CPU (calls `std::vector<xla::Literal> literals = runtime::GetComputationClient()->TransferFromDevice(UnwrapXlaData(device_data));`).